[dtensor] refactor and improve readability of _dispatch.py

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132914
* #132683
* __->__ #132682
* #131210

as titled. It also changes some comments of _op_schema.py to make them
update to date

cc @XilunWu @H-Huang @awgu @kwen2501 @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

wanchaol committed:
[dtensor] refactor and improve readability of _dispatch.py

as titled. It also changes some comments of _op_schema.py to make them
update to date

[ghstack-poisoned]

XilunWu commented on torch/distributed/_tensor/_dispatch.py:
```diff
@@ -418,3 +386,36 @@ def wrap(res: object, spec: OutputSpecType) -> object:
+                "or explicitly create a DTensor under distributed enviroment."
+            )
+
+        if tensor_arg.numel() == 1 or self._allow_implicit_replication:
+            # scalar tensor can be safely treated as replicated
+            replication_spec = DTensorSpec(
+                mesh,
+                (Replicate(),) * mesh.ndim,
+                tensor_meta=TensorMeta(
+                    shape=tensor_arg.shape,
+                    stride=tensor_arg.stride(),
+                    dtype=tensor_arg.dtype,
+                ),
+            )
```
@mori360 this is where the `inf` and `scale` arguments are converted to `Replicate()` DTenstor in the op you were examining last week.

wanchaol committed:
Update on "[dtensor] refactor and improve readability of _dispatch.py"


as titled. It also changes some comments of _op_schema.py to make them
update to date

cc XilunWu H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]
